File: (03-22-2024)Product of Array Except Self/joshlin/solution.ts
function productExceptSelf(nums: number[]): number[] {
    let left = 1;
    let right = 1;
    const res = Array<number>(nums.length).fill(1);
    for (let i = 1; i < nums.length; i++) {
        left *= nums[i - 1];
        res[i] = left;
    };
    for (let i = nums.length - 2; i >= 0; i--) {
        right  *= nums[i + 1];
        res[i] *= right;
    };
    return res;
};

File: (03-22-2024)Product of Array Except Self/joshlin/solution.py

from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # [1, 2, 3, 4]
        #  1  1  2  6  left to right
        #  24 12 4  1  right to left
        #  24 12 8  6  return list
        res = [1] * len(nums)
        left, right = 1, 1

        for i in range(1, len(nums)):
            left *= nums[i - 1]
            res[i] = left
        
        for i in range(len(nums)-2, -1, -1):
            right  *= nums[i + 1]
            res[i] *= right

        return res